### PR TITLE
Support get local image by containerd

### PR DESCRIPTION
**Pull request by mowangdk**
Fix https://github.com/aquasecurity/trivy/issues/851

done:
code develop
e2e test 
image

Current state: 
add unit test

### CODE REVIEW

**Comment by mowangdk** [comment]
@knqyf263 e2e test passed,  i need write unit test for containerd daemon, should i add code on this project? https://github.com/aquasecurity/testdocker

**Comment by knqyf263** [comment]
@mowangdk Yes, it would be great.

**Comment by mowangdk** [comment]
@knqyf263 i wonder if there is any other ways to complete auto tests such as add test case in e2e tests? Because containerd official repo does not provide any mock interfaces. there are many interfaces need to mock which requires a lot of work

**Comment by mowangdk** [comment]
@knqyf263 I'll keep writing the mock interface if there's no other way, but it takes a long time。。

**Comment by knqyf263** [comment]
Can we emulate containerd only for necessary APIs?

**Comment by mowangdk** [comment]
> Can we emulate containerd only for necessary APIs?

@knqyf263  I was intent to do this, but containerd is designed with a complex interface and it is incompatible with the docker. So i am going to create an interface on the client side to wrap all the methods provided by containerd. mock these method on the new interface

**Comment by knqyf263** [comment]
Is there any tool supporting containerd? We may want to see how those tools test communication with containerd.

**Comment by mowangdk** [comment]
> Is there any tool supporting containerd? We may want to see how those tools test communication with containerd.

I have checked [cri](https://github.com/containerd/containerd/tree/main/pkg/cri) and [pouch](https://github.com/alibaba/pouch/blob/master/ctrd/wrapper_client.go). Neither project provides an containerd openapi for simulation.

**Comment by knqyf263** [comment]
How do they test containerd?

**Comment by mowangdk** [comment]
> How do they test containerd?
@knqyf263 pouch itself does not provide containerd related unittest, and cri does so in the way I mentioned above~

**Comment by knqyf263** [comment]
What if we give up unit tests and focus on integration tests? We may want to run containerd as a container as follows.
https://github.com/westonsteimel/container-library-containerd/pkgs/container/containerd

**Comment by mowangdk** [comment]
> What if we give up unit tests and focus on integration tests? We may want to run containerd as a container as follows. https://github.com/westonsteimel/container-library-containerd/pkgs/container/containerd
> 
> Also, you can refer to this Dockerfile. https://github.com/containerd/nerdctl/blob/1e0afeb8067dd85d18b7b2d3554dd71aba45dc02/Dockerfile#L52-L66

gotcha, I think this is a better way, I will refer to the ```integration/``` examples to complete the integration test

**Comment by knqyf263** [comment]
Thanks for your efforts.

**Comment by mowangdk** [comment]
> Thanks for your efforts.

u're welcome~

**Comment by SebastianBakala-TomTom** [comment]
Hello Guys.
Any update on that? :)

**Comment by mowangdk** [comment]
> Hello Guys. Any update on that? :)

not finish yet, in the process of writing integration tests

**Comment by mowangdk** [comment]
@knqyf263 I have encountered a new problem, please advise how to solve. i can't access the socket path outside the container (the path is already shared)

docker inspect result:
image

no response from host access
image

This issue prevents me from using testcontainer to access containerd，containerd will throws this error
image

**Comment by knqyf263** [comment]
Did you try Privileged=true?
https://pkg.go.dev/github.com/testcontainers/testcontainers-go#ContainerRequest

**Comment by mowangdk** [comment]
> Did you try Privileged=true? https://pkg.go.dev/github.com/testcontainers/testcontainers-go#ContainerRequest

Sorry for the late reply, i'm still in vacation. yes, I have tried privileged params. but it's still failed. i will try to ask this question on slack

docker inspect info:
```json
[
    {
        "Id": "xxxx",
        "Created": "xxxx",
        "Path": "/bin/sh",
        "Args": [
            "-c",
            "mkdir -p /etc/containerd/ \u0026\u0026 /usr/local/bin/containerd config default \u003e /etc/containerd/config.toml \u0026\u0026 /usr/local/bin/containerd -c /etc/containerd/config.toml"
        ],
        "State": {
            "Status": "running",
            "Running": true,
            "Paused": false,
            "Restarting": false,
            "OOMKilled": false,
            "Dead": false,
            "Pid": 2650,
            "ExitCode": 0,
            "Error": "",
            "StartedAt": "2022-02-04T12:22:41.2548137Z",
            "FinishedAt": "0001-01-01T00:00:00Z"
        },
        "Image": "sha256:eeff5658ca7390778dc69bc2fba1f119c303876430becb9cf8249bae85151dac",
        "ResolvConfPath": "/var/lib/docker/containers/aaee1af1c6267d6c4e0b166cabfd2ee1045074c46c37ac843aeb84a75fb22f5f/resolv.conf",
        "HostnamePath": "/var/lib/docker/containers/aaee1af1c6267d6c4e0b166cabfd2ee1045074c46c37ac843aeb84a75fb22f5f/hostname",
        "HostsPath": "/var/lib/docker/containers/aaee1af1c6267d6c4e0b166cabfd2ee1045074c46c37ac843aeb84a75fb22f5f/hosts",
        "LogPath": "/var/lib/docker/containers/aaee1af1c6267d6c4e0b166cabfd2ee1045074c46c37ac843aeb84a75fb22f5f/aaee1af1c6267d6c4e0b166cabfd2ee1045074c46c37ac843aeb84a75fb22f5f-json.log",
        "Name": "/containerd",
        "RestartCount": 0,
        "Driver": "overlay2",
        "Platform": "linux",
        "MountLabel": "",
        "ProcessLabel": "",
        "AppArmorProfile": "",
        "ExecIDs": null,
        "HostConfig": {
            "Binds": null,
            "ContainerIDFile": "",
            "LogConfig": {
                "Type": "json-file",
                "Config": {}
            },
            "NetworkMode": "default",
            "PortBindings": {},
            "RestartPolicy": {
                "Name": "",
                "MaximumRetryCount": 0
            },
            "AutoRemove": false,
            "VolumeDriver": "",
            "VolumesFrom": null,
            "CapAdd": null,
            "CapDrop": null,
            "CgroupnsMode": "host",
            "Dns": null,
            "DnsOptions": null,
            "DnsSearch": null,
            "ExtraHosts": null,
            "GroupAdd": null,
            "IpcMode": "private",
            "Cgroup": "",
            "Links": null,
            "OomScoreAdj": 0,
            "PidMode": "",
            "Privileged": true,
            "PublishAllPorts": false,
            "ReadonlyRootfs": false,
            "SecurityOpt": [
                "label=disable"
            ],
            "UTSMode": "",
            "UsernsMode": "",
            "ShmSize": 67108864,
            "Runtime": "runc",
            "ConsoleSize": [
                0,
                0
            ],
            "Isolation": "",
            "CpuShares": 0,
            "Memory": 0,
            "NanoCpus": 0,
            "CgroupParent": "",
            "BlkioWeight": 0,
            "BlkioWeightDevice": null,
            "BlkioDeviceReadBps": null,
            "BlkioDeviceWriteBps": null,
            "BlkioDeviceReadIOps": null,
            "BlkioDeviceWriteIOps": null,
            "CpuPeriod": 0,
            "CpuQuota": 0,
            "CpuRealtimePeriod": 0,
            "CpuRealtimeRuntime": 0,
            "CpusetCpus": "",
            "CpusetMems": "",
            "Devices": null,
            "DeviceCgroupRules": null,
            "DeviceRequests": null,
            "KernelMemory": 0,
            "KernelMemoryTCP": 0,
            "MemoryReservation": 0,
            "MemorySwap": 0,
            "MemorySwappiness": null,
            "OomKillDisable": false,
            "PidsLimit": null,
            "Ulimits": null,
            "CpuCount": 0,
            "CpuPercent": 0,
            "IOMaximumIOps": 0,
            "IOMaximumBandwidth": 0,
            "Mounts": [
                {
                    "Type": "bind",
                    "Source": "/Users/katsueiki/test/test",
                    "Target": "/run/containerd"
                }
            ],
            "MaskedPaths": null,
            "ReadonlyPaths": null
        },
        "GraphDriver": {
            "Data": {
                "LowerDir": "/var/lib/docker/overlay2/2f5838d77dbe4a0e715e0a3b6b4f6a4a52cc9473768f450947bce73b8ee54347-init/diff:/var/lib/docker/overlay2/5ukmzhl8grghfmlb037fqgkup/diff:/var/lib/docker/overlay2/jrhhonombshwj98dbz62oy7oe/diff:/var/lib/docker/overlay2/6d3f0742f9518e2fec05c85f239d570b9bc7b91661d970f0ae90c4d87221a998/diff:/var/lib/docker/overlay2/a5e27e2d2390bba820777fba664f9519b18b81a652f12232b542d93fc03a3319/diff",
                "MergedDir": "/var/lib/docker/overlay2/2f5838d77dbe4a0e715e0a3b6b4f6a4a52cc9473768f450947bce73b8ee54347/merged",
                "UpperDir": "/var/lib/docker/overlay2/2f5838d77dbe4a0e715e0a3b6b4f6a4a52cc9473768f450947bce73b8ee54347/diff",
                "WorkDir": "/var/lib/docker/overlay2/2f5838d77dbe4a0e715e0a3b6b4f6a4a52cc9473768f450947bce73b8ee54347/work"
            },
            "Name": "overlay2"
        },
        "Mounts": [
            {
                "Type": "bind",
                "Source": "/Users/katsueiki/test/test",
                "Destination": "/run/containerd",
                "Mode": "",
                "RW": true,
                "Propagation": "rprivate"
            }
        ],
        "Config": {
            "Hostname": "aaee1af1c626",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
                "CONTAINERD_VERSION=main",
                "CONTAINERD_REVISION=9c676e98dde995db46841a2c6126ad709e3fc7d7"
            ],
            "Cmd": null,
            "Image": "docker.io/geyingqi/dockercontainerd:alpine",
            "Volumes": null,
            "WorkingDir": "",
            "Entrypoint": [
                "/bin/sh",
                "-c",
                "mkdir -p /etc/containerd/ \u0026\u0026 /usr/local/bin/containerd config default \u003e /etc/containerd/config.toml \u0026\u0026 /usr/local/bin/containerd -c /etc/containerd/config.toml"
            ],
            "OnBuild": null,
            "Labels": {
                "org.opencontainers.image.description": "containerd in Docker",
                "org.opencontainers.image.revision": "9c676e98dde995db46841a2c6126ad709e3fc7d7",
                "org.opencontainers.image.title": "containerd",
                "org.opencontainers.image.version": "main",
                "org.testcontainers.golang": "true",
                "org.testcontainers.golang.sessionId": "8e18df93-51ca-48ca-8a21-f82287609e1d"
            }
        },
        "NetworkSettings": {
            "Bridge": "",
            "SandboxID": "d6f50c778904da8f2cc7c11ba714256cbf6244d5f1bb35a83e8e5baa9599c0ca",
            "HairpinMode": false,
            "LinkLocalIPv6Address": "",
            "LinkLocalIPv6PrefixLen": 0,
            "Ports": {},
            "SandboxKey": "/var/run/docker/netns/d6f50c778904",
            "SecondaryIPAddresses": null,
            "SecondaryIPv6Addresses": null,
            "EndpointID": "",
            "Gateway": "",
            "GlobalIPv6Address": "",
            "GlobalIPv6PrefixLen": 0,
            "IPAddress": "",
            "IPPrefixLen": 0,
            "IPv6Gateway": "",
            "MacAddress": "",
            "Networks": {
                "host": {
                    "IPAMConfig": null,
                    "Links": null,
                    "Aliases": null,
                    "NetworkID": "144cedd89c10cd96cc8775b8310e48e3346a38f6ceefe0a881ee1aed8281f04c",
                    "EndpointID": "2a9f10d93b55cfacba62178651688e3be3593dc459c33be0c83242d9eefa2266",
                    "Gateway": "",
                    "IPAddress": "",
                    "IPPrefixLen": 0,
                    "IPv6Gateway": "",
                    "GlobalIPv6Address": "",
                    "GlobalIPv6PrefixLen": 0,
                    "MacAddress": "",
                    "DriverOpts": null
                }
            }
        }
    }
]

```

**Comment by mowangdk** [comment]
@knqyf263 added some [unit tests](https://github.com/aquasecurity/fanal/pull/348/files#diff-2aba0fe9e63d1c40578ff5a4fd435d44f8069a107d346f0d02d455ab36ffaa73R17-R22) by mock containerd interface, please check if they meet the requirements. If it is OK I will continue to improve these unit tests first. (The integration test issue is still being confirmed, so I can't guarantee that it will be solved)

**Comment by mowangdk** [comment]
@knqyf263 integration test is ready, It seems to be unable to connect to the socket file on mac, switch to linux and it's OK

**Comment by knqyf263** [comment]
Thanks for the great contribution! It is really helpful. This PR is large and we have many tasks currently. I'm sorry it will take a while to review it.

**Comment by mowangdk** [comment]
> I have triggered. It would be better to run tests in your local first.

okay, thx~

**Comment by mowangdk** [comment]
@knqyf263 
image
image

The unit test running on my vm is successful, but the unit test in ci reported a permission error, has this been encountered?

**Comment by AndreyLevchenko** [comment]
Hi @mowangdk 
could you please resolve merge conflict?

**Comment by mowangdk** [comment]
> Hi @mowangdk could you please resolve merge conflict?

integration test still failed, could you help me to find out the reason?  I haven't changed the testTLSRegistry part of the code.
image

**Comment by AndreyLevchenko** [comment]
Hi @mowangdk 
Let me investigate issue with integration tests

**Comment by mowangdk** [comment]
@AndreyLevchenko All tests have passed, thanks for your help, can you tell me what is the reason?

**Comment by AndreyLevchenko** [comment]
> @AndreyLevchenko All tests have passed, thanks for your help, can you tell me what is the reason?
Hi
Sorry for delayed answer
There were several issues:
- socket permissions
- `BindMounts` map requires swap of key and value
- missed `WaitingFor` property
- Several `testcontainers` invocations while integration tests (I just borrowed config from `trivy`)

File: image/daemon/containerd.go [hunk 1]
```diff
@@ -30,7 +30,7 @@ func imageWriter(client *containerd.Client, img containerd.Image) imageSave {
 		}
 		imgOpts := archive.WithImage(client.ImageService(), ref[0])
 		manifestOpts := archive.WithManifest(img.Target())
-		platOpts := archive.WithPlatform(platforms.Default())
+		platOpts := archive.WithPlatform(platforms.OnlyStrict(platforms.DefaultSpec()))
```
Comment (knqyf263) [inline image/daemon/containerd.go:33]: Can we use about DefaultStrict()?
https://github.com/containerd/containerd/blob/ebcf876bf7b8f1cb79c60f37af08affa135c398d/platforms/defaults.go#L25

Comment (AndreyLevchenko) [reply image/daemon/containerd.go]: yes we can

**Comment by mowangdk** [comment]
> * WaitingFor

Seems that the reason is very complicated... Anyway, thanks for your help!

File: test/integration/containerd_test.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	require.NoError(t, err)
+	hostPath := filepath.Join(normalPath, "data", "test")
+	targetPath := filepath.Join(hostPath, "containerd")
+	defer os.RemoveAll(hostPath)
+	err = os.MkdirAll(targetPath, os.ModePerm)
+	require.NoError(t, err)
+
+	req := testcontainers.ContainerRequest{
+		Name:       "containerd",
+		Image:      "docker.io/geyingqi/dockercontainerd:alpine",
```
Comment (knqyf263) [inline test/integration/containerd_test.go:41]: We should not depend on the 3rd-party image. What if we put Dockerfile in this repository and push the image to ghcr.io/aquasecurity?

File: test/integration/containerd_test.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/stretchr/testify/require"
+	testcontainers "github.com/testcontainers/testcontainers-go"
+	"github.com/testcontainers/testcontainers-go/wait"
+)
+
+const (
+	redisImageName  = "index.docker.io/library/redis:latest"
+	mysqlImageName  = "index.docker.io/library/mysql:latest"
+	pythonImageName = "index.docker.io/library/python:3.4-alpine"
```
Comment (knqyf263) [inline test/integration/containerd_test.go:24]: Can we import tarballs into containerd?
https://github.com/aquasecurity/fanal/blob/d15bba5c3713716a99826c8500a260536f322fad/Makefile#L19

That is what we do for Docker Engine.
https://github.com/aquasecurity/fanal/blob/d15bba5c3713716a99826c8500a260536f322fad/test/integration/library_test.go#L171-L174

File: test/integration/containerd_test.go (hunk 2 above)
Comment (knqyf263) [reply test/integration/containerd_test.go:41]: If you will upload Dockerfile, I will build an image and push it to GHCR. Thanks.

File: image/daemon.go [hunk 4]
```diff
@@ -1,10 +1,18 @@
...
 	"github.com/google/go-containerregistry/pkg/name"
 
 	"github.com/aquasecurity/fanal/image/daemon"
 	"github.com/aquasecurity/fanal/types"
+	"github.com/containerd/containerd/namespaces"
+)
+
+const (
+	defaultContainerdSocket    = "/run/containerd/containerd.sock"
+	defaultContainerdNamespace = "default"
```
Comment (knqyf263) [inline image/daemon.go:15]: These values should be in `containerd.go`.

**Comment by mowangdk** [comment]
Sorry, I've been busy with work lately.  unable to follow up in time 😭

**Review by knqyf263 — APPROVED** [review]

**Comment by knqyf263** [comment]
@mowangdk Thank you so much!

**Pull request merged**